presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,321 @@
+    {
+        return new ComparisonExpression(
+                node.getOperator(),
+                (Expression) process(node.getLeft(), context),
+                (Expression) process(node.getRight(), context));
+    }
+
+    @Override
+    protected Node visitLiteral(Literal node, MaterializedViewQueryOptimizerContext context)
+    {
+        return node;
+    }
+
+    @Override
+    protected Node visitGroupBy(GroupBy node, MaterializedViewQueryOptimizerContext context)
+    {
+        ImmutableList.Builder<GroupingElement> rewrittenGroupBy = ImmutableList.builder();
+        for (GroupingElement element : node.getGroupingElements()) {
+            GroupingElement rewrittenElement = (GroupingElement) process(element, context);
+            rewrittenGroupBy.add(rewrittenElement);
+        }
+        return new GroupBy(
+                node.isDistinct(),
+                rewrittenGroupBy.build());
+    }
+
+    @Override
+    protected Node visitOrderBy(OrderBy node, MaterializedViewQueryOptimizerContext context)
+    {
+        ImmutableList.Builder<SortItem> rewrittenOrderBy = ImmutableList.builder();
+        for (SortItem sortItem : node.getSortItems()) {
+            SortItem rewrittenSortItem = (SortItem) process(sortItem, context);
+            rewrittenOrderBy.add(rewrittenSortItem);
+        }
+        return new OrderBy(rewrittenOrderBy.build());
+    }
+
+    @Override
+    protected Node visitSortItem(SortItem node, MaterializedViewQueryOptimizerContext context)
+    {
+        return new SortItem(
+                (Expression) process(node.getSortKey(), context),
+                node.getOrdering(),
+                node.getNullOrdering());
+    }
+
+    @Override
+    protected Node visitSimpleGroupBy(SimpleGroupBy node, MaterializedViewQueryOptimizerContext context)
+    {
+        ImmutableList.Builder<Expression> rewrittenSimpleGroupBy = ImmutableList.builder();
+        for (Expression column : node.getExpressions()) {
+            Expression rewrittenColumn = (Expression) process(column, context);
+            rewrittenSimpleGroupBy.add(rewrittenColumn);
+        }
+        return new SimpleGroupBy(rewrittenSimpleGroupBy.build());
+    }
+
+    public static final class MaterializedViewQueryOptimizerContext
+    {
+        private final Table materializedView;
+        private final ImmutableMap<String, String> baseToViewColumnMap;
+        private final Optional<Relation> baseQueryRelation;
+        private boolean rewritable;
+
+        public MaterializedViewQueryOptimizerContext(
+                Table materializedView,
+                Query originalSqlQuery)
+        {
+            this.materializedView = requireNonNull(materializedView, "materialized view is null");
+
+            QuerySpecification originalSqlQueryBody = (QuerySpecification) originalSqlQuery.getQueryBody();
+            this.baseQueryRelation = originalSqlQueryBody.getFrom().isPresent() ? originalSqlQueryBody.getFrom() : Optional.empty();
```

Comment:
Do you want to call it materializedViewDefinition or so? I took me a while to realize it is not the user input query.